### PR TITLE
Proxy for Unreal Insights data streamed from server workers

### PR DESCRIPTION
**Pull request by improbable-danny**
#### Description
This script adds support for using a proxy connection to stream Unreal Insights data from a cloud worker. 

Engine PR: https://github.com/improbableio/UnrealEngine/pull/848/files

#### Primary reviewers
@m-samiec

### CODE REVIEW

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            worker_connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            worker_connection.connect(("127.0.0.1", 1981))
+            worker_connection.settimeout(1.0)
+            is_connected = True
+        except socket.error:
+            pass
+            
+    print("Connecting to 127.0.0.1:1980 [Insights]")
+    proxy_connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    proxy_connection.connect(("127.0.0.1", 1980))
```
Comment (simonsarginson) [inline SpatialGDK/Extras/insights-proxy/InsightsProxy.py:18]: minor: worth putting in to variable and using `f'Connecting to {up}:{port} [Insights]'` format strings?

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+def main():
+    is_connected = False
+    while not is_connected:
+        try:
+            print("Attempting connection to 127.0.0.1:1981 [server worker]")
+            worker_connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            worker_connection.connect(("127.0.0.1", 1981))
+            worker_connection.settimeout(1.0)
+            is_connected = True
+        except socket.error:
```
Comment (simonsarginson) [inline SpatialGDK/Extras/insights-proxy/InsightsProxy.py:13]: Do we need some timeout with an exception in here so this doesn't spin forever?

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    proxy_connection.connect(("127.0.0.1", 1980))
+    
+    print("Waiting for target..")
+
+    while True:
+        try:
+            chunk = worker_connection.recv(16*1024*1024)
+        except socket.timeout:
+            continue
+        print "Forwarding " + str(len(chunk))
```
Comment (simonsarginson) [inline SpatialGDK/Extras/insights-proxy/InsightsProxy.py:27]: would prefer print() syntax so we are forwards compatible with python 3

File: SpatialGDK/Extras/insights-proxy/README.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# Insights Proxy
+This Python script is used to allow proxying insights data from a SpatialOS worker to Insights running on the local machine. See UNR-5207 for more information.
```
Comment (simonsarginson) [inline SpatialGDK/Extras/insights-proxy/README.md:2]: So is this directly invoked on a dev machine? This is python 2.x code at the moment, would be good to describe this dependency, would also prefer to be python 3.6+ based personally

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py (hunk 1 above)
Comment (simonsarginson) [reply SpatialGDK/Extras/insights-proxy/InsightsProxy.py:18]: Would need to upgrade to python 3.6 to do that. Python 3.0 has Connecting to `'{}:{} [Insights]'.format(ip, port)` which achieves the same, or if sticking with 2.x c style %s format is available

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+def main():
+    is_connected = False
+    while not is_connected:
+        try:
+            print("Attempting connection to 127.0.0.1:1981 [server worker]")
+            worker_connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            worker_connection.connect(("127.0.0.1", 1981))
+            worker_connection.settimeout(1.0)
+            is_connected = True
```
Comment (simonsarginson) [inline SpatialGDK/Extras/insights-proxy/InsightsProxy.py:12]: very minor: could just break here and us while True, save you a variable

File: SpatialGDK/Extras/insights-proxy/README.md (hunk 4 above)
Comment (improbable-danny) [reply SpatialGDK/Extras/insights-proxy/README.md:2]: Yeah, and run manually. Good point about Python versions, is it possible to make code compatible with as many (all?) versions as possible?

Comment (simonsarginson) [reply SpatialGDK/Extras/insights-proxy/README.md:2]: 3.6 allows for format string syntax. But I don't think I got who the target audience was until I dived into the docs just now. If you want wide support use `print()` other than that your code should work in python 2 & 3. If you want to add the string formatting suggestion I commented above you'll need to restrict yourself to c style string formatting: e.g. `print('Connecting to %s:%d [Insights]' % (ip, port))`

Comment (simonsarginson) [reply SpatialGDK/Extras/insights-proxy/README.md:2]: It's worth noting that unreal comes packaged with python 3.7 (turned on by default) and 2.7 (turned off by default)

Comment (simonsarginson) [reply SpatialGDK/Extras/insights-proxy/README.md]: I think the workflow doc describes a workflow that depends on whatever python you have installed on PATH

Comment (simonsarginson) [reply SpatialGDK/Extras/insights-proxy/README.md]: So actually you could run this from unreal directly? https://docs.unrealengine.com/en-US/ProductionPipelines/ScriptingAndAutomation/Python/index.html

**Review by simonsarginson — APPROVED** [review]
LGTM

**Comment by m-samiec** [comment]
Please add experimental note, and changelog

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import socket 
+import sys 
+
+TARGET_IP = "127.0.0.1"
+TARGET_PORT = 1981
```
Comment (m-samiec) [inline SpatialGDK/Extras/insights-proxy/InsightsProxy.py:5]: add comment linking dependent port no in engine

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py [hunk 7]
```diff
@@ -1,21 +1,24 @@
...
-            worker_connection.connect(("127.0.0.1", 1981))
+            worker_connection.connect((TARGET_IP, TARGET_PORT))
             worker_connection.settimeout(1.0)
-            is_connected = True
+            break
         except socket.error:
             pass
             
-    print("Connecting to 127.0.0.1:1980 [Insights]")
+    print("Connecting to the running Insights instance 127.0.0.1:%s" % INSIGHTS_PORT)
```
Comment (simonsarginson) [inline SpatialGDK/Extras/insights-proxy/InsightsProxy.py:19]: minor: use TARGET_IP

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py [hunk 8]
```diff
@@ -1,21 +1,24 @@
...
     proxy_connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    proxy_connection.connect(("127.0.0.1", 1980))
+    proxy_connection.connect(("127.0.0.1", INSIGHTS_PORT))
```
Comment (simonsarginson) [inline SpatialGDK/Extras/insights-proxy/InsightsProxy.py:21]: minor: use TARGET_IP

File: SpatialGDK/Extras/insights-proxy/InsightsProxy.py (hunk 2 above)
Comment (improbable-danny) [reply SpatialGDK/Extras/insights-proxy/InsightsProxy.py:19]: I think spinning should be ok, this is a manual process and I sort of leave it up to the developer to ensure the port is accessible and listening (triggered from Insights) - what are your thoughts?

**Pull request merged**